File: scripts/fetch_commits.js
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

function runGitLog(workspace, author, since, until) {
    if (!fs.existsSync(workspace) || !fs.statSync(workspace).isDirectory()) {
        return [];
    }

    const gitDir = path.join(workspace, '.git');
    if (!fs.existsSync(gitDir)) {
        return [];
    }

    const cmdArgs = ['git', 'log', '--no-merges', '--pretty=format:%s'];
    if (author) cmdArgs.push('--author', `"${author}"`);
    if (since) cmdArgs.push('--since', `"${since}"`);
    if (until) cmdArgs.push('--until', `"${until}"`);

    try {
        const output = execSync(cmdArgs.join(' '), { cwd: workspace, stdio: ['pipe', 'pipe', 'pipe'] }).toString().trim();
        if (!output) return [];
        return output.split('\n').map(line => line.trim()).filter(line => line);
    } catch (e) {
        return [];
    }
}

function main() {
    const args = process.argv.slice(2);
    let author = '', since = '', until = '', output = '';
    const workspaces = [];

    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--author') author = args[++i];
        else if (args[i] === '--since') since = args[++i];
        else if (args[i] === '--until') until = args[++i];
        else if (args[i] === '--output') output = args[++i];
        else if (args[i] === '--workspaces') {
            i++;
            while (i < args.length && !args[i].startsWith('--')) {
                workspaces.push(args[i]);
                i++;
            }
            i--;
        }
    }

    if (!output || workspaces.length === 0) {
        console.error('Error: --output and --workspaces are required.');
        process.exit(1);
    }

    const report = {};

    for (const workspace of workspaces) {
        const projectName = path.basename(path.resolve(workspace));
        const commits = runGitLog(workspace, author, since, until);
        if (commits.length > 0) {
            report[projectName] = commits;
        }
    }

    try {
        fs.writeFileSync(output, JSON.stringify(report, null, 2), 'utf8');
        console.log(`Success! Data written to: ${output}`);
    } catch (e) {
        console.error(`Error writing to output file: ${e.message}`);
        process.exit(1);
    }
}

main();

File: scripts/fetch_commits.py
import argparse
import subprocess
import json
import os
import sys

def run_git_log(workspace, author, since, until):
    """Run git log in the specified workspace and return a list of commit messages."""
    if not os.path.isdir(workspace):
        return []
    
    # Check if it's a git repo
    git_dir = os.path.join(workspace, '.git')
    if not os.path.exists(git_dir):
        return []

    cmd = ['git', 'log', '--no-merges', '--pretty=format:%s']
    
    if author:
        cmd.extend(['--author', author])
    if since:
        cmd.extend(['--since', since])
    if until:
        cmd.extend(['--until', until])

    try:
        result = subprocess.run(cmd, cwd=workspace, capture_output=True, text=True, check=True)
        output = result.stdout.strip()
        if not output:
            return []
        
        # Split by newline and filter empty lines
        commits = [line.strip() for line in output.split('\n') if line.strip()]
        return commits
    except subprocess.CalledProcessError as e:
        # Silently catch git command failures (e.g. not a git repo, no commits)
        return []

def main():
    parser = argparse.ArgumentParser(description="Fetch git commits across multiple workspaces.")
    parser.add_argument('--author', type=str, help="Filter by author name or email (e.g., 'John Doe' or 'john@example.com')")
    parser.add_argument('--since', type=str, help="Start date/time (e.g., 'midnight', '2023-10-01')")
    parser.add_argument('--until', type=str, help="End date/time (e.g., 'now', '2023-10-31')")
    parser.add_argument('--workspaces', nargs='+', required=True, help="List of absolute paths to workspace directories")
    parser.add_argument('--output', type=str, required=True, help="Output JSON file path")
    
    args = parser.parse_args()
    
    report = {}
    
    for workspace in args.workspaces:
        # Use the folder name as the project identifier
        project_name = os.path.basename(os.path.normpath(workspace))
        commits = run_git_log(workspace, args.author, args.since, args.until)
        if commits:
            report[project_name] = commits
    
    # Write to output file
    try:
        with open(args.output, 'w', encoding='utf-8') as f:
            json.dump(report, f, indent=2, ensure_ascii=False)
        print(f"Success! Data written to: {args.output}")
    except Exception as e:
        print(f"Error writing to output file: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == '__main__':
    main()
